hy/scoping.py

```diff
@@ -0,0 +1,243 @@
+            self.exposing_assignments = True
+
+        return ret
+
+    def __exit__(self, *args):
+        for node in self.name_nodes:
+            if node.name not in self.assignments:
+                # pass unbound/nonlocal names up to parent scope
+                self.parent.access(node)
+        return super().__exit__(*args)
+
+    def _assign(self, node):
+        if node.name in self.iterators:
+            raise SyntaxError("assignment expression cannot rebind comprehension iteration variable")
+        if node.name not in self.assignments:
+            self.assignments.append(node.name)
+        self.name_nodes.append(node)
+
+    def _access(self, node):
+        self.name_nodes.append(node)
+
+    def iterator(self, target):
+        for name in ast.walk(target):
+            if isinstance(name, ast.Name):
+                self.iterators.add(name.id)
+        self.assignments = [n for n in self.assignments
+                            if n not in self.iterators]
+        self.name_nodes = [node for node in self.name_nodes
+                           if node.name not in self.iterators]
+
+
+# Used for fns, defns, and defclasses
+class ScopeFn(Scope):
+    def __init__(self, stack, scope_id, args=None):
+        super().__init__(stack, scope_id)
+        self.defined = set()
+        self.name_nodes = []
+        self.nonlocal_vars = set()
+        self.is_fn = args is not None
+
+        if args:
+            for arg in itertools.chain(args.args,
+                                       args.posonlyargs, args.kwonlyargs,
+                                       [args.vararg, args.kwarg]):
+                if arg:
+                    self.define(arg.arg)
+
+    def __exit__(self, *args):
+        for node in self.name_nodes:
+            if node.name not in self.defined or node.name in self.nonlocal_vars:
+                # pass unbound/nonlocal names up to parent scope
+                self.parent.access(node)
+        return super().__exit__(*args)
+
+    def _assign(self, node):
+        self.access(node)
+        self.define(node.name)
+
+    def _access(self, node):
+        self.name_nodes.append(node)
+
+    def define(self, name):
+        self.defined.add(name)
+
+    def unlocal(self, nonlocal_node):
+        # should consider allowing nonlocals in other scopes
+        self.nonlocal_vars.update(nonlocal_node.names)
+        for node in self.name_nodes:
+            if node.name in nonlocal_node.names:
+                raise SyntaxError(f"name `{node.name}` is declared nonlocal after being used")
+        # toss all nonlocal names up to parent scope
+        if isinstance(nonlocal_node, ast.Nonlocal):
+            for i in range(len(nonlocal_node.names)):
+                self.parent.access(nonlocal_node, i)
+
+
+class ScopeStack:
+    """Scoping information that Hy adds on top of Python's scoping rules."""
+    def __init__(self):
+        self.last_id = 0
+        self.scope = Scope(self, self.last_id)
```

Comment:
no other code instantiates `Scope` directly. would it be more appropriate to make it a true abstract base class inheriting from `abc.ABC` and adding a `ScopeRoot` class that is instantiated here instead?